File: runtime/store.js
import { constants as fsConstants } from "node:fs";
import { access, mkdir, open, readFile, rename, rm, stat } from "node:fs/promises";
import path from "node:path";
import { setTimeout as delay } from "node:timers/promises";
import { ProtocolError, clone, parseJsonStrict, record, validateAnalysis } from "./orchestrator.js";
import { applyEvent } from "./events.js";
import { parseLegacyPlan, renderPlan } from "./render.js";
import { reserveNext } from "./routing.js";
import { newState, validateState } from "./state.js";
function isWithin(base, candidate) {
    const relative = path.relative(path.resolve(base), path.resolve(candidate));
    return relative === "" || (!relative.startsWith(`..${path.sep}`) && relative !== ".." && !path.isAbsolute(relative));
}
async function exists(candidate) {
    try {
        await access(candidate, fsConstants.F_OK);
        return true;
    }
    catch {
        return false;
    }
}
async function atomicWrite(candidate, content) {
    await mkdir(path.dirname(candidate), { recursive: true });
    const temporary = path.join(path.dirname(candidate), `.${path.basename(candidate)}.${process.pid}.${Date.now()}.tmp`);
    const handle = await open(temporary, "wx", 0o600);
    try {
        await handle.writeFile(content, "utf8");
        await handle.sync();
    }
    finally {
        await handle.close();
    }
    try {
        await rename(temporary, candidate);
    }
    catch (error) {
        const code = error.code;
        if (!new Set(["EEXIST", "EPERM", "EACCES"]).has(code ?? ""))
            throw error;
        await rm(candidate, { force: true });
        await rename(temporary, candidate);
    }
    finally {
        await rm(temporary, { force: true });
    }
}
async function parseJsonFile(candidate) {
    let content;
    try {
        content = await readFile(candidate, "utf8");
    }
    catch (error) {
        throw new ProtocolError(candidate, "cannot read JSON", String(error));
    }
    try {
        return parseJsonStrict(content);
    }
    catch (error) {
        if (error instanceof ProtocolError)
            throw error;
        throw new ProtocolError(candidate, "invalid JSON", String(error));
    }
}
async function appendJournal(candidate, entry) {
    const entries = [];
    if (await exists(candidate)) {
        for (const [index, line] of (await readFile(candidate, "utf8")).split(/\r?\n/).filter(Boolean).entries()) {
            try {
                entries.push(record(parseJsonStrict(line), `journal[${index}]`));
            }
            catch (error) {
                throw new ProtocolError(`journal[${index}]`, "invalid JSON", String(error));
            }
        }
    }
    if (entries.some((item) => item.entry_id === entry.entry_id))
        return;
    entries.push(entry);
    await atomicWrite(candidate, entries.map((item) => JSON.stringify(item)).join("\n") + "\n");
}
export class WorkflowStore {
    base;
    root;
    internal;
    statePath;
    planPath;
    analysisPath;
    journalPath;
    transactionPath;
    lockPath;
    request;
    constructor(directory, request) {
        newState(request);
        this.base = path.resolve(directory);
        this.root = path.resolve(this.base, "1_orchestrator", request);
        const expectedParent = path.resolve(this.base, "1_orchestrator");
        if (path.dirname(this.root) !== expectedParent)
            throw new ProtocolError("workflow_root", "request path escapes workflow base", this.root);
        this.request = request;
        this.internal = path.join(this.root, ".orchestrator");
        this.statePath = path.join(this.internal, "state.json");
        this.planPath = path.join(this.root, "plan.md");
        this.analysisPath = path.join(this.root, "analysis.json");
        this.journalPath = path.join(this.internal, "journal.jsonl");
        this.transactionPath = path.join(this.internal, "transaction.json");
        this.lockPath = path.join(this.internal, "lock");
    }
    async ensureRoot() {
        await mkdir(this.internal, { recursive: true });
        const resolved = path.resolve(this.root);
        if (path.dirname(resolved) !== path.resolve(this.base, "1_orchestrator"))
            throw new ProtocolError("workflow_root", "resolved request path escapes workflow base");
    }
    async withLock(operation, timeoutMs = 5000, staleMs = 300000) {
        await this.ensureRoot();
        const deadline = Date.now() + timeoutMs;
        while (true) {
            try {
                const handle = await open(this.lockPath, "wx", 0o600);
                await handle.writeFile(JSON.stringify({ pid: process.pid, created_at: new Date().toISOString() }));
                await handle.close();
                break;
            }
            catch (error) {
                if (error.code !== "EEXIST")
                    throw error;
                try {
                    if (Date.now() - (await stat(this.lockPath)).mtimeMs > staleMs) {
                        await rm(this.lockPath, { force: true });
                        continue;
                    }
                }
                catch {
                    continue;
                }
                if (Date.now() >= deadline)
                    throw new ProtocolError("workflow_lock", "request is already being advanced");
                await delay(25);
            }
        }
        try {
            return await operation();
        }
        finally {
            await rm(this.lockPath, { force: true });
        }
    }
    async recover() {
        if (!(await exists(this.transactionPath)))
            return false;
        const transaction = record(await parseJsonFile(this.transactionPath), "transaction");
        if (transaction.schema_version !== 1)
            throw new ProtocolError("transaction.schema_version", "unsupported transaction");
        await atomicWrite(this.statePath, `${JSON.stringify(transaction.state, null, 2)}\n`);
        if (typeof transaction.plan !== "string" || !transaction.plan.trim())
            throw new ProtocolError("transaction.plan", "must be a non-empty string");
        await atomicWrite(this.planPath, transaction.plan);
        await appendJournal(this.journalPath, record(transaction.journal, "transaction.journal"));
        await rm(this.transactionPath, { force: true });
        return true;
    }
    async loadState() {
        await this.ensureRoot();
        await this.recover();
        if (await exists(this.statePath))
            return validateState(await parseJsonFile(this.statePath));
        if (await exists(this.planPath))
            return parseLegacyPlan(await readFile(this.planPath, "utf8"), this.request);
        return newState(this.request);
    }
    async loadAnalysis() {
        return await exists(this.analysisPath) ? validateAnalysis(await parseJsonFile(this.analysisPath)) : undefined;
    }
    journal(action, state, detail) {
        const transition = typeof detail.transition_id === "string" ? detail.transition_id : "state";
        return {
            entry_id: `${transition}:${action}:${state.state_revision}`,
            timestamp: new Date().toISOString(),
            action,
            state_revision: state.state_revision,
            transition_id: detail.transition_id ?? null,
            detail: clone(detail),
        };
    }
    async commit(state, analysis, journal) {
        const validated = validateState(state, analysis && state.stages.length && !state.legacy_migrated ? analysis : undefined);
        const plan = renderPlan(validated, analysis);
        const transaction = { schema_version: 1, state: validated, plan, journal };
        await atomicWrite(this.transactionPath, `${JSON.stringify(transaction, null, 2)}\n`);
        await atomicWrite(this.statePath, `${JSON.stringify(validated, null, 2)}\n`);
        await atomicWrite(this.planPath, plan);
        await appendJournal(this.journalPath, journal);
        await rm(this.transactionPath, { force: true });
    }
    async reserve(expectedStateRevision) {
        return this.withLock(async () => {
            const state = await this.loadState();
            const analysis = await this.loadAnalysis();
            const result = reserveNext(state, analysis, expectedStateRevision);
            if (JSON.stringify(result.state) !== JSON.stringify(state))
                await this.commit(result.state, analysis, this.journal("reserve", result.state, result.action));
            return result;
        });
    }
    async apply(event, expectedStateRevision) {
        return this.withLock(async () => {
            const state = await this.loadState();
            const analysis = await this.loadAnalysis();
            const result = await applyEvent(this.base, state, event, analysis, expectedStateRevision);
            if (JSON.stringify(result.state) !== JSON.stringify(state))
                await this.commit(result.state, analysis, this.journal("apply", result.state, { transition_id: event.transition_id, event_type: event.type, result: result.result }));
            return result;
        });
    }
    async validate() {
        return this.withLock(async () => {
            const state = await this.loadState();
            const analysis = await this.loadAnalysis();
            validateState(state, analysis && state.stages.length && !state.legacy_migrated ? analysis : undefined);
            if (state.legacy_migrated && !(await exists(this.statePath))) {
                await this.commit(state, analysis, this.journal("migrate", state, { transition_id: null, source: "legacy-plan.md" }));
            }
            const issues = [];
            const expected = renderPlan(state, analysis);
            if (await exists(this.planPath) && await readFile(this.planPath, "utf8") !== expected)
                issues.push("plan.md differs from deterministic rendering");
            if (["review", "reviewed", "approved"].includes(state.analysis_status) && !analysis)
                issues.push("analysis.json is required by current state");
            return { valid: !issues.length, state_revision: state.state_revision, status: state.status, pending: state.pending, issues };
        });
    }
}

File: runtime/orchestrator_core/store.py
from __future__ import annotations

from copy import deepcopy
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Mapping

from .events import apply_event
from .io import append_jsonl_once, atomic_write, exclusive_lock, json_bytes, recover_transaction
from .migration import legacy_to_state
from .model import REQUEST_ID, new_state, validate_state
from .protocol import ProtocolError, load_json
from .render import render_plan
from .routing import reserve_next
from .traceability import validate_execution_graph


class WorkflowStore:
    def __init__(self, workflow_base: Path, request_id: str):
        if REQUEST_ID.fullmatch(request_id) is None:
            raise ProtocolError("request_id", "invalid request identifier", request_id)
        self.workflow_base = workflow_base.expanduser().resolve()
        self.root = (self.workflow_base / "1_orchestrator" / request_id).resolve()
        if self.root.parent != (self.workflow_base / "1_orchestrator").resolve():
            raise ProtocolError("workflow_root", "request path escapes workflow base", str(self.root))
        self.request_id = request_id
        self.internal = self.root / ".orchestrator"
        self.state_path = self.internal / "state.json"
        self.journal_path = self.internal / "journal.jsonl"
        self.transaction_path = self.internal / "transaction.json"
        self.lock_path = self.internal / "lock"
        self.plan_path = self.root / "plan.md"
        self.analysis_path = self.root / "analysis.json"

    def _ensure_root(self) -> None:
        self.internal.mkdir(parents=True, exist_ok=True)
        if self.root.resolve().parent != (self.workflow_base / "1_orchestrator").resolve():
            raise ProtocolError("workflow_root", "resolved request path escapes workflow base")

    def lock(self, *, timeout: float = 5.0, stale_after: float = 300.0):
        self._ensure_root()
        return exclusive_lock(self.lock_path, timeout=timeout, stale_after=stale_after)

    def recover(self) -> bool:
        self._ensure_root()
        return recover_transaction(self.transaction_path, self.state_path, self.plan_path, self.journal_path)

    def load_state(self) -> dict[str, Any]:
        self._ensure_root()
        self.recover()
        if self.state_path.exists():
            return validate_state(load_json(self.state_path))
        if self.plan_path.exists():
            backup = self.internal / "legacy-plan.md"
            if not backup.exists():
                atomic_write(backup, self.plan_path.read_bytes())
            return legacy_to_state(self.plan_path, self.request_id)
        return new_state(self.request_id)

    def load_analysis(self) -> dict[str, Any] | None:
        return validate_execution_graph(load_json(self.analysis_path)) if self.analysis_path.exists() else None

    def _journal(self, action: str, state: Mapping[str, Any], detail: Mapping[str, Any]) -> dict[str, Any]:
        transition_id = detail.get("transition_id")
        return {
            "entry_id": f"{transition_id or 'state'}:{action}:{state['state_revision']}",
            "timestamp": datetime.now(timezone.utc).isoformat(), "action": action,
            "state_revision": state["state_revision"], "transition_id": transition_id,
            "detail": deepcopy(dict(detail)),
        }

    def _commit(self, state: Mapping[str, Any], analysis: Mapping[str, Any] | None, journal: Mapping[str, Any]) -> None:
        cross_check = analysis if analysis is not None and state.get("stages") and not state.get("legacy_migrated") else None
        validated = validate_state(state, cross_check)
        plan = render_plan(validated, analysis)
        transaction = {"schema_version": 1, "state": validated, "plan": plan, "journal": dict(journal)}
        atomic_write(self.transaction_path, json_bytes(transaction))
        atomic_write(self.state_path, json_bytes(validated))
        atomic_write(self.plan_path, plan.encode(), 0o644)
        append_jsonl_once(self.journal_path, journal)
        self.transaction_path.unlink(missing_ok=True)

    def reserve(self, *, expected_state_revision: int | None = None) -> tuple[dict[str, Any], dict[str, Any]]:
        with self.lock():
            state, analysis = self.load_state(), self.load_analysis()
            next_state, action = reserve_next(state, analysis, expected_state_revision=expected_state_revision)
            if next_state != state:
                self._commit(next_state, analysis, self._journal("reserve", next_state, action))
            return next_state, action

    def apply(self, event: Mapping[str, Any], *, expected_state_revision: int | None = None) -> tuple[dict[str, Any], dict[str, Any]]:
        with self.lock():
            state, analysis = self.load_state(), self.load_analysis()
            next_state, result = apply_event(state, event, analysis, expected_state_revision=expected_state_revision)
            if next_state != state:
                detail = {"transition_id": event.get("transition_id"), "event_type": event.get("type"), "result": result}
                self._commit(next_state, analysis, self._journal("apply", next_state, detail))
            return next_state, result

    def validate(self) -> dict[str, Any]:
        with self.lock():
            state, analysis = self.load_state(), self.load_analysis()
            cross_check = analysis if analysis is not None and state["stages"] and not state["legacy_migrated"] else None
            validate_state(state, cross_check)
            issues: list[str] = []
            expected = render_plan(state, analysis)
            if self.plan_path.exists() and self.plan_path.read_text(encoding="utf-8") != expected:
                issues.append("plan.md differs from deterministic rendering")
            if state["analysis_status"] in {"review", "reviewed", "approved"} and analysis is None:
                issues.append("analysis.json is required by current state")
            return {"valid": not issues, "state_revision": state["state_revision"], "status": state["status"], "pending": state["pending"], "issues": issues}
